src/compiler/checker.ts

```diff
@@ -16204,24 +16252,67 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+                                    indexInfos.push(createIndexInfo(keyType, declaration.type ? getTypeFromTypeNode(declaration.type) : anyType, hasEffectiveModifier(declaration, ModifierFlags.Readonly), declaration));
+                                }
+                            });
+                        }
+                    }
+                }
+                else if (hasLateBindableIndexSignature(declaration)) {
+                    const declName = isBinaryExpression(declaration) ? declaration.left as ElementAccessExpression : (declaration as LateBoundDeclaration).name;
+                    const keyType = isElementAccessExpression(declName) ? checkExpressionCached(declName.argumentExpression) : checkComputedPropertyName(declName);
+                    if (findIndexInfo(indexInfos, keyType)) {
+                        continue; // Explicit index for key type takes priority
+                    }
+                    if (isTypeAssignableTo(keyType, stringNumberSymbolType)) {
+                        if (isTypeAssignableTo(keyType, numberType)) {
+                            hasComputedNumberProperty = true;
+                            if (!hasEffectiveReadonlyModifier(declaration)) {
+                                readonlyComputedNumberProperty = false;
                             }
-                        });
+                        }
+                        else if (isTypeAssignableTo(keyType, esSymbolType)) {
+                            hasComputedSymbolProperty = true;
+                            if (!hasEffectiveReadonlyModifier(declaration)) {
+                                readonlyComputedSymbolProperty = false;
+                            }
+                        }
+                        else {
+                            hasComputedStringProperty = true;
+                            if (!hasEffectiveReadonlyModifier(declaration)) {
+                                readonlyComputedStringProperty = false;
+                            }
+                        }
+                        computedPropertySymbols.push(declaration.symbol);
                     }
                 }
             }
+            const allPropertySymbols = concatenate(computedPropertySymbols, filter(siblingSymbols, s => s !== indexSymbol));
+            // aggregate similar index infos implied to be the same key to the same combined index info
+            if (hasComputedStringProperty && !findIndexInfo(indexInfos, stringType)) indexInfos.push(getObjectLiteralIndexInfo(readonlyComputedStringProperty, 0, allPropertySymbols, stringType));
+            if (hasComputedNumberProperty && !findIndexInfo(indexInfos, numberType)) indexInfos.push(getObjectLiteralIndexInfo(readonlyComputedNumberProperty, 0, allPropertySymbols, numberType));
+            if (hasComputedSymbolProperty && !findIndexInfo(indexInfos, esSymbolType)) indexInfos.push(getObjectLiteralIndexInfo(readonlyComputedSymbolProperty, 0, allPropertySymbols, esSymbolType));

```

Comment:
Are all of the `findIndexInfo` calls likely to cost a bunch or is `indexInfos` just always short?

Reply:
In practice, it's always short (there's usually just `string` or `number` indexes) - in theory, a user could explicitly define hundreds of pattern literal index infos and it'd be bad, however, and you'd prefer you did the up-front work to make a map keyed on key type (if you had multiple kinds of computed names). But that's not code anyone's realistically writing.